prog/vm/nexus.rb

```diff
@@ -162,16 +159,30 @@ def storage_secrets
   def allocation_dataset
     requires_bdev_ubi = frame["storage_volumes"].any? { |v| v["use_bdev_ubi"] }
     spdk_where_clause = requires_bdev_ubi ? "AND id in (select vm_host_id from spdk_installation where version like '%ubi%' and allocation_weight > 0)" : ""
-    DB[<<SQL, vm.cores, vm.mem_gib_ratio, vm.mem_gib, vm_storage_size_gib, vm.location, vm.arch]
+
+    # YYY: Although when the following WHERE clauses return true then disks can
+    # be allocated, but there are edge cases where disks can be allocated, but
+    # the following WHERE clauses return false. Improve them.
+    max_storage_gib = frame["storage_volumes"].map { |v| v["size_gib"] }.max
+    total_storage_gib = frame["storage_volumes"].sum { |v| v["size_gib"] }
+    n_volumes = frame["storage_volumes"].length
+    device_where_clause = frame["distinct_storage_devices"] ?
+        "AND (SELECT count(*) FROM storage_device WHERE " \
+          "storage_device.enabled AND storage_device.vm_host_id = vm_host.id AND " \
+          "storage_device.available_storage_gib >= #{max_storage_gib}) >= #{n_volumes}"
+        : "AND (SELECT max(available_storage_gib) FROM storage_device WHERE " \
+        "storage_device.enabled AND storage_device.vm_host_id = vm_host.id) >= #{total_storage_gib}"
+
```

Comment:
@byucesoy @furkansahin @enescakir, can you work through this query and some of the supporting data structures/stack entities? I would do so myself, but it's getting late for me and the sun rises for you next. Also, @furkansahin might have some special insight into how he'd prefer to program against this, to direct how devices map between the VM and the host.

Reply:
Basic idea is:
* If `distinct_storage_devices` is false, then we find a host which has a device which can store all disks.
* If `distinct_storage_devices` is true, then we find a host which has n_volumes devices which can store the largest disk.

Although the above doesn't cover all allocatable cases, but when it returns true, it means we can allocate. And it was easy to express as SQL, so I chose to do it as the first step. We can improve on this in follow-up PRs.